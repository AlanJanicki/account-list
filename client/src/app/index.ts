export * from './services';
export { setupStore } from './store/store';
