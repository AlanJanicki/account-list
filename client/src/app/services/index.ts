export * from './accountList';
export { api } from './api';
export * from './routes';
