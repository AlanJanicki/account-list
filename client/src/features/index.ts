export { List as AccountList, HEADERS } from './AccountList/List';
