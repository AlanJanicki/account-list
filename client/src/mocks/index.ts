export * from './accountList';
export { default as accounts } from './accountList/accounts.json';
export { default as accountTypes } from './accountList/accounttypes.json';
export * from './msw';
