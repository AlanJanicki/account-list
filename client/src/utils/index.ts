export * from './styles/';
export { transformToArrayOfObjectsWithGivenKeys } from './table';
