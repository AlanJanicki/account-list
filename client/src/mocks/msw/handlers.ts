import { ResponseComposition, rest, RestContext, RestRequest } from 'msw';

import { restApiRoutes } from '../../app/';
import { default as accountsMock } from '../accountList/accounts.json';
import { default as accountTypesMock } from '../accountList/accounttypes.json';

const { accountTypes, accounts } = restApiRoutes;

export const httpErrors = {
  accounts: rest.get(accounts, (_, res, ctx) => res(ctx.status(500)))
};

export const responseOKWithPayload = (
  _: RestRequest,
  res: ResponseComposition,
  ctx: RestContext,
  payload: Record<string, unknown>[]
) => res(ctx.status(200), ctx.json(payload));

export const handlers = [
  rest.get(accounts, (...args) => responseOKWithPayload(...args, accountsMock)),
  rest.get(accountTypes, (...args) => responseOKWithPayload(...args, accountTypesMock))
];
