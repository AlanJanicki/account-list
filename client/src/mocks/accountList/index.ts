import { v4 as uuidv4 } from 'uuid';

import { HEADERS } from '../../features';
import { transformToArrayOfObjectsWithGivenKeys } from '../../utils';
import accounts from './accounts.json';

export const getCellsWithMockedData = () => {
  const headers = Object.values(HEADERS).map(({ key }) => key);

  return [
    ...headers.map((value) => ({ isHeader: true, key: uuidv4(), value })),
    ...transformToArrayOfObjectsWithGivenKeys({
      data: accounts.slice(0, 3),
      keys: headers,
      valueFallback: '-'
    })
  ];
};
