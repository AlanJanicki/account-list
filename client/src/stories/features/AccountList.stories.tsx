import { Meta, StoryFn } from '@storybook/react';
import { mswDecorator } from 'msw-storybook-addon';
import React from 'react';
import { Provider } from 'react-redux';

import { setupStore } from '../../app/';
import { AccountList } from '../../features';
import { handlers, httpErrors } from '../../mocks/msw/handlers';

export default {
  component: AccountList,
  parameters: {
    msw: {
      handlers,
    },
  },
  title: 'AccountList',
};

const Template = {
  decorators: [
    (Story: StoryFn) => (
      <Provider store={setupStore()}>
        <Story />
      </Provider>
    ),
    mswDecorator,
  ],
} as Meta;

export const Default = { ...Template };

export const HttpError = { ...Template };
HttpError.parameters = {
  msw: {
    handlers: [httpErrors.accounts],
  },
};
