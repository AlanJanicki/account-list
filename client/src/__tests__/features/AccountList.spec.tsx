import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { rest } from 'msw';
import React from 'react';

import { restApiRoutes } from '../../app/';
import { AccountList, HEADERS } from '../../features';
import { accounts, accountTypes, httpErrors, responseOKWithPayload, server } from '../../mocks';
import { transformToArrayOfObjectsWithGivenKeys } from '../../utils';
import { matchSnapshot, renderWithStore } from '../common';

type InvalidData = Record<string, unknown>[];

const getCells = async () =>
  (await screen.findAllByTestId('cell')).map(({ textContent }) => textContent);

const getAccountListToDisplay = (accountList: typeof accounts) =>
  accountList.map(({ accountType, currency, name, profitLoss }) => ({
    accountType: accountTypes.find(({ id }) => id === accountType)?.title ?? '',
    name,
    profitLoss: !(`${profitLoss ?? ''}` || undefined) ? '' : `${currency} ${profitLoss}`
  }));

export const getDataToDisplay = (accountList = accounts) =>
  transformToArrayOfObjectsWithGivenKeys({
    data: getAccountListToDisplay(accountList),
    keys: Object.values(HEADERS).map(({ key }) => key),
    valueFallback: '-'
  }).map(({ value }) => value);

const getAccountsWithInvalidData = (
  keys: string[],
  invalidValue?: null | string | number
): InvalidData => [
  {
    ...Object.entries(accounts[0]).reduce((acc, [key, value]) => {
      const changeKey = keys.some((property) => property === key);
      return { ...acc, [key]: changeKey ? invalidValue : value };
    }, {})
  },
  ...accounts.slice(1)
];

const testDisplayingCellsWithInvalidData = async (invalidData: InvalidData) => {
  mockServerResponse(invalidData);
  const { asFragment } = renderWithStore(<AccountList />);
  expect(await getCells()).toEqual(getDataToDisplay(invalidData as typeof accounts));
  matchSnapshot(asFragment());
};

const mockServerResponse = (payload: InvalidData) =>
  server.use(
    rest.get(restApiRoutes.accounts, (...args) => responseOKWithPayload(...args, payload))
  );

describe('<AccountList/> - layout', () => {
  it('displays spinner during api call', async () => {
    const { asFragment } = renderWithStore(<AccountList />);
    matchSnapshot(asFragment());
    await screen.findByRole('status');
  });

  it('hides spinner on api call finish', async () => {
    const { asFragment } = renderWithStore(<AccountList />);
    await waitFor(() => expect(screen.queryByRole('status')).not.toBeInTheDocument());
    matchSnapshot(asFragment());
  });

  it('displays card header', async () => {
    renderWithStore(<AccountList />);
    await screen.findByRole('heading', { name: 'Account list' });
  });

  it('displays headers', async () => {
    renderWithStore(<AccountList />);
    const headers = (await screen.findAllByTestId('header')).map(({ textContent }) => textContent);
    expect(Object.values(HEADERS).map(({ value }) => value)).toEqual(headers);
  });

  it('displays cells with valid data', async () => {
    renderWithStore(<AccountList />);
    expect(await getCells()).toEqual(getDataToDisplay());
  });

  it(`displays "No accounts to display" when accountList is empty`, async () => {
    mockServerResponse([]);
    const { asFragment } = renderWithStore(<AccountList />);
    expect(await screen.findByText('No accounts to display')).toBeInTheDocument();
    matchSnapshot(asFragment());
  });

  it.each(
    [undefined, null, ''].map((value) => ({
      name: typeof value === 'string' ? 'empty' : value,
      value
    }))
  )(`displays fallback '-' when account has $name values`, async ({ value }) => {
    await testDisplayingCellsWithInvalidData(
      getAccountsWithInvalidData(
        Object.values(HEADERS).map(({ key }) => key),
        value
      )
    );
  });

  it(`displays fallback '-' when profitLoss is 0`, async () => {
    await testDisplayingCellsWithInvalidData(getAccountsWithInvalidData(['profitLoss'], 0));
  });
});

describe('<AccountList/> - server error', () => {
  beforeEach(() => server.use(httpErrors.accounts));

  const getRetryButton = async () => await screen.findByRole('button', { name: 'Try again' });

  it('displays error message on accountList fetch failure', async () => {
    const { asFragment } = renderWithStore(<AccountList />);
    await screen.findByText('Something went wrong fetching account list');
    matchSnapshot(asFragment());
  });

  it('displays refetch button on accountList fetch failure', async () => {
    renderWithStore(<AccountList />);
    await getRetryButton();
  });

  it('displays spinner during refetch api call', async () => {
    const { asFragment } = renderWithStore(<AccountList />);
    userEvent.click(await getRetryButton());
    matchSnapshot(asFragment());
    await screen.findByRole('status');
  });

  it('hides spinner on refetch api call finish', async () => {
    renderWithStore(<AccountList />);
    userEvent.click(await getRetryButton());
    await waitFor(() => expect(screen.queryByRole('status')).not.toBeInTheDocument());
  });
});
