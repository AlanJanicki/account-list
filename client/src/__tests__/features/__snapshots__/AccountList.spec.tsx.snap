// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AccountList/> - layout displays "No accounts to display" when accountList is empty 1`] = `
<DocumentFragment>
  .c1 {
  margin: 0 auto;
  width: 80%;
  border: 1px solid #dfdfdf;
  border-radius: .25rem;
}

.c2 {
  height: 3rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #f7f7f7;
}

.c3 {
  padding: 1rem;
  display: grid;
}

.c0 {
  padding-top: 2rem;
}

.c4 {
  text-align: center;
}

@media (min-width:576px) {
  .c2 {
    height: 3.5rem;
  }
}

@media (min-width:576px) {
  .c2 h1 {
    font-size: 1.1rem;
  }
}

@media (min-width:768px) {
  .c2 h1 {
    font-size: 1.2rem;
  }
}

@media (min-width:992px) {
  .c2 h1 {
    font-size: 1.5rem;
  }
}

@media (min-width:992px) {
  .c4 {
    font-size: 1.1rem;
  }
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <h1>
          Account list
        </h1>
      </div>
      <div
        class="c3"
      >
        <p
          class="c4"
        >
          No accounts to display
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<AccountList/> - layout displays fallback '-' when account has empty values 1`] = `
<DocumentFragment>
  .c1 {
  margin: 0 auto;
  width: 80%;
  border: 1px solid #dfdfdf;
  border-radius: .25rem;
}

.c2 {
  height: 3rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #f7f7f7;
}

.c3 {
  padding: 1rem;
  display: grid;
}

.c4 {
  display: grid;
  grid-template-columns: repeat(3,1fr);
  overflow: scroll;
}

.c4 span {
  padding: 1rem 0.3rem;
  border-bottom: 1px solid #dfdfdf;
}

.c4 span:nth-child(-n + 3) {
  border-bottom: 1px solid #000;
}

.c4 span:nth-last-child(-n + 3) {
  border-bottom: none;
}

.c0 {
  padding-top: 2rem;
}

@media (min-width:576px) {
  .c2 {
    height: 3.5rem;
  }
}

@media (min-width:576px) {
  .c2 h1 {
    font-size: 1.1rem;
  }
}

@media (min-width:768px) {
  .c2 h1 {
    font-size: 1.2rem;
  }
}

@media (min-width:992px) {
  .c2 h1 {
    font-size: 1.5rem;
  }
}

@media (min-width:768px) {
  .c4 span {
    padding: 1.2rem 0.3rem;
  }
}

@media (min-width:992px) {
  .c4 span {
    font-size: 1.1rem;
  }
}

@media (min-width:768px) {
  .c4 span:nth-child(-n + 3) > * {
    font-size: 1.1rem;
  }
}

@media (min-width:992px) {
  .c4 span:nth-child(-n + 3) > * {
    font-size: 1.3rem;
  }
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <h1>
          Account list
        </h1>
      </div>
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <span
            data-testid="header"
          >
            <strong>
              Name
            </strong>
          </span>
          <span
            data-testid="header"
          >
            <strong>
              Profit & Loss
            </strong>
          </span>
          <span
            data-testid="header"
          >
            <strong>
              Account Type
            </strong>
          </span>
          <span
            data-testid="cell"
          >
            -
          </span>
          <span
            data-testid="cell"
          >
            -
          </span>
          <span
            data-testid="cell"
          >
            -
          </span>
          <span
            data-testid="cell"
          >
            New Spread bet
          </span>
          <span
            data-testid="cell"
          >
            $ -679
          </span>
          <span
            data-testid="cell"
          >
            Spread bet account
          </span>
          <span
            data-testid="cell"
          >
            Spread bet - demo
          </span>
          <span
            data-testid="cell"
          >
            $ 16.211
          </span>
          <span
            data-testid="cell"
          >
            Spread bet account
          </span>
          <span
            data-testid="cell"
          >
            CFD - demo
          </span>
          <span
            data-testid="cell"
          >
            € 16679
          </span>
          <span
            data-testid="cell"
          >
            CFD account
          </span>
          <span
            data-testid="cell"
          >
            Stockbroking - DEMO
          </span>
          <span
            data-testid="cell"
          >
            $ 200045
          </span>
          <span
            data-testid="cell"
          >
            Share dealing account
          </span>
          <span
            data-testid="cell"
          >
            Stockbroking
          </span>
          <span
            data-testid="cell"
          >
            € 200045
          </span>
          <span
            data-testid="cell"
          >
            Share dealing account
          </span>
          <span
            data-testid="cell"
          >
            My CFD
          </span>
          <span
            data-testid="cell"
          >
            $ -10
          </span>
          <span
            data-testid="cell"
          >
            CFD account
          </span>
          <span
            data-testid="cell"
          >
            Forex Acccount
          </span>
          <span
            data-testid="cell"
          >
            $ 10
          </span>
          <span
            data-testid="cell"
          >
            Forex account
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<AccountList/> - layout displays fallback '-' when account has null values 1`] = `
<DocumentFragment>
  .c1 {
  margin: 0 auto;
  width: 80%;
  border: 1px solid #dfdfdf;
  border-radius: .25rem;
}

.c2 {
  height: 3rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #f7f7f7;
}

.c3 {
  padding: 1rem;
  display: grid;
}

.c4 {
  display: grid;
  grid-template-columns: repeat(3,1fr);
  overflow: scroll;
}

.c4 span {
  padding: 1rem 0.3rem;
  border-bottom: 1px solid #dfdfdf;
}

.c4 span:nth-child(-n + 3) {
  border-bottom: 1px solid #000;
}

.c4 span:nth-last-child(-n + 3) {
  border-bottom: none;
}

.c0 {
  padding-top: 2rem;
}

@media (min-width:576px) {
  .c2 {
    height: 3.5rem;
  }
}

@media (min-width:576px) {
  .c2 h1 {
    font-size: 1.1rem;
  }
}

@media (min-width:768px) {
  .c2 h1 {
    font-size: 1.2rem;
  }
}

@media (min-width:992px) {
  .c2 h1 {
    font-size: 1.5rem;
  }
}

@media (min-width:768px) {
  .c4 span {
    padding: 1.2rem 0.3rem;
  }
}

@media (min-width:992px) {
  .c4 span {
    font-size: 1.1rem;
  }
}

@media (min-width:768px) {
  .c4 span:nth-child(-n + 3) > * {
    font-size: 1.1rem;
  }
}

@media (min-width:992px) {
  .c4 span:nth-child(-n + 3) > * {
    font-size: 1.3rem;
  }
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <h1>
          Account list
        </h1>
      </div>
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <span
            data-testid="header"
          >
            <strong>
              Name
            </strong>
          </span>
          <span
            data-testid="header"
          >
            <strong>
              Profit & Loss
            </strong>
          </span>
          <span
            data-testid="header"
          >
            <strong>
              Account Type
            </strong>
          </span>
          <span
            data-testid="cell"
          >
            -
          </span>
          <span
            data-testid="cell"
          >
            -
          </span>
          <span
            data-testid="cell"
          >
            -
          </span>
          <span
            data-testid="cell"
          >
            New Spread bet
          </span>
          <span
            data-testid="cell"
          >
            $ -679
          </span>
          <span
            data-testid="cell"
          >
            Spread bet account
          </span>
          <span
            data-testid="cell"
          >
            Spread bet - demo
          </span>
          <span
            data-testid="cell"
          >
            $ 16.211
          </span>
          <span
            data-testid="cell"
          >
            Spread bet account
          </span>
          <span
            data-testid="cell"
          >
            CFD - demo
          </span>
          <span
            data-testid="cell"
          >
            € 16679
          </span>
          <span
            data-testid="cell"
          >
            CFD account
          </span>
          <span
            data-testid="cell"
          >
            Stockbroking - DEMO
          </span>
          <span
            data-testid="cell"
          >
            $ 200045
          </span>
          <span
            data-testid="cell"
          >
            Share dealing account
          </span>
          <span
            data-testid="cell"
          >
            Stockbroking
          </span>
          <span
            data-testid="cell"
          >
            € 200045
          </span>
          <span
            data-testid="cell"
          >
            Share dealing account
          </span>
          <span
            data-testid="cell"
          >
            My CFD
          </span>
          <span
            data-testid="cell"
          >
            $ -10
          </span>
          <span
            data-testid="cell"
          >
            CFD account
          </span>
          <span
            data-testid="cell"
          >
            Forex Acccount
          </span>
          <span
            data-testid="cell"
          >
            $ 10
          </span>
          <span
            data-testid="cell"
          >
            Forex account
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<AccountList/> - layout displays fallback '-' when account has undefined values 1`] = `
<DocumentFragment>
  .c1 {
  margin: 0 auto;
  width: 80%;
  border: 1px solid #dfdfdf;
  border-radius: .25rem;
}

.c2 {
  height: 3rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #f7f7f7;
}

.c3 {
  padding: 1rem;
  display: grid;
}

.c4 {
  display: grid;
  grid-template-columns: repeat(3,1fr);
  overflow: scroll;
}

.c4 span {
  padding: 1rem 0.3rem;
  border-bottom: 1px solid #dfdfdf;
}

.c4 span:nth-child(-n + 3) {
  border-bottom: 1px solid #000;
}

.c4 span:nth-last-child(-n + 3) {
  border-bottom: none;
}

.c0 {
  padding-top: 2rem;
}

@media (min-width:576px) {
  .c2 {
    height: 3.5rem;
  }
}

@media (min-width:576px) {
  .c2 h1 {
    font-size: 1.1rem;
  }
}

@media (min-width:768px) {
  .c2 h1 {
    font-size: 1.2rem;
  }
}

@media (min-width:992px) {
  .c2 h1 {
    font-size: 1.5rem;
  }
}

@media (min-width:768px) {
  .c4 span {
    padding: 1.2rem 0.3rem;
  }
}

@media (min-width:992px) {
  .c4 span {
    font-size: 1.1rem;
  }
}

@media (min-width:768px) {
  .c4 span:nth-child(-n + 3) > * {
    font-size: 1.1rem;
  }
}

@media (min-width:992px) {
  .c4 span:nth-child(-n + 3) > * {
    font-size: 1.3rem;
  }
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <h1>
          Account list
        </h1>
      </div>
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <span
            data-testid="header"
          >
            <strong>
              Name
            </strong>
          </span>
          <span
            data-testid="header"
          >
            <strong>
              Profit & Loss
            </strong>
          </span>
          <span
            data-testid="header"
          >
            <strong>
              Account Type
            </strong>
          </span>
          <span
            data-testid="cell"
          >
            -
          </span>
          <span
            data-testid="cell"
          >
            -
          </span>
          <span
            data-testid="cell"
          >
            -
          </span>
          <span
            data-testid="cell"
          >
            New Spread bet
          </span>
          <span
            data-testid="cell"
          >
            $ -679
          </span>
          <span
            data-testid="cell"
          >
            Spread bet account
          </span>
          <span
            data-testid="cell"
          >
            Spread bet - demo
          </span>
          <span
            data-testid="cell"
          >
            $ 16.211
          </span>
          <span
            data-testid="cell"
          >
            Spread bet account
          </span>
          <span
            data-testid="cell"
          >
            CFD - demo
          </span>
          <span
            data-testid="cell"
          >
            € 16679
          </span>
          <span
            data-testid="cell"
          >
            CFD account
          </span>
          <span
            data-testid="cell"
          >
            Stockbroking - DEMO
          </span>
          <span
            data-testid="cell"
          >
            $ 200045
          </span>
          <span
            data-testid="cell"
          >
            Share dealing account
          </span>
          <span
            data-testid="cell"
          >
            Stockbroking
          </span>
          <span
            data-testid="cell"
          >
            € 200045
          </span>
          <span
            data-testid="cell"
          >
            Share dealing account
          </span>
          <span
            data-testid="cell"
          >
            My CFD
          </span>
          <span
            data-testid="cell"
          >
            $ -10
          </span>
          <span
            data-testid="cell"
          >
            CFD account
          </span>
          <span
            data-testid="cell"
          >
            Forex Acccount
          </span>
          <span
            data-testid="cell"
          >
            $ 10
          </span>
          <span
            data-testid="cell"
          >
            Forex account
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<AccountList/> - layout displays fallback '-' when profitLoss is 0 1`] = `
<DocumentFragment>
  .c1 {
  margin: 0 auto;
  width: 80%;
  border: 1px solid #dfdfdf;
  border-radius: .25rem;
}

.c2 {
  height: 3rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #f7f7f7;
}

.c3 {
  padding: 1rem;
  display: grid;
}

.c4 {
  display: grid;
  grid-template-columns: repeat(3,1fr);
  overflow: scroll;
}

.c4 span {
  padding: 1rem 0.3rem;
  border-bottom: 1px solid #dfdfdf;
}

.c4 span:nth-child(-n + 3) {
  border-bottom: 1px solid #000;
}

.c4 span:nth-last-child(-n + 3) {
  border-bottom: none;
}

.c0 {
  padding-top: 2rem;
}

@media (min-width:576px) {
  .c2 {
    height: 3.5rem;
  }
}

@media (min-width:576px) {
  .c2 h1 {
    font-size: 1.1rem;
  }
}

@media (min-width:768px) {
  .c2 h1 {
    font-size: 1.2rem;
  }
}

@media (min-width:992px) {
  .c2 h1 {
    font-size: 1.5rem;
  }
}

@media (min-width:768px) {
  .c4 span {
    padding: 1.2rem 0.3rem;
  }
}

@media (min-width:992px) {
  .c4 span {
    font-size: 1.1rem;
  }
}

@media (min-width:768px) {
  .c4 span:nth-child(-n + 3) > * {
    font-size: 1.1rem;
  }
}

@media (min-width:992px) {
  .c4 span:nth-child(-n + 3) > * {
    font-size: 1.3rem;
  }
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <h1>
          Account list
        </h1>
      </div>
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <span
            data-testid="header"
          >
            <strong>
              Name
            </strong>
          </span>
          <span
            data-testid="header"
          >
            <strong>
              Profit & Loss
            </strong>
          </span>
          <span
            data-testid="header"
          >
            <strong>
              Account Type
            </strong>
          </span>
          <span
            data-testid="cell"
          >
            Spread bet
          </span>
          <span
            data-testid="cell"
          >
            $ 0
          </span>
          <span
            data-testid="cell"
          >
            Spread bet account
          </span>
          <span
            data-testid="cell"
          >
            New Spread bet
          </span>
          <span
            data-testid="cell"
          >
            $ -679
          </span>
          <span
            data-testid="cell"
          >
            Spread bet account
          </span>
          <span
            data-testid="cell"
          >
            Spread bet - demo
          </span>
          <span
            data-testid="cell"
          >
            $ 16.211
          </span>
          <span
            data-testid="cell"
          >
            Spread bet account
          </span>
          <span
            data-testid="cell"
          >
            CFD - demo
          </span>
          <span
            data-testid="cell"
          >
            € 16679
          </span>
          <span
            data-testid="cell"
          >
            CFD account
          </span>
          <span
            data-testid="cell"
          >
            Stockbroking - DEMO
          </span>
          <span
            data-testid="cell"
          >
            $ 200045
          </span>
          <span
            data-testid="cell"
          >
            Share dealing account
          </span>
          <span
            data-testid="cell"
          >
            Stockbroking
          </span>
          <span
            data-testid="cell"
          >
            € 200045
          </span>
          <span
            data-testid="cell"
          >
            Share dealing account
          </span>
          <span
            data-testid="cell"
          >
            My CFD
          </span>
          <span
            data-testid="cell"
          >
            $ -10
          </span>
          <span
            data-testid="cell"
          >
            CFD account
          </span>
          <span
            data-testid="cell"
          >
            Forex Acccount
          </span>
          <span
            data-testid="cell"
          >
            $ 10
          </span>
          <span
            data-testid="cell"
          >
            Forex account
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<AccountList/> - layout displays spinner during api call 1`] = `
<DocumentFragment>
  .c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  text-align: center;
}

.c2 {
  width: 50px;
  display: block;
  aspect-ratio: 1;
  border: 3px solid #0f7bff;
  border-radius: 50%;
  border-top-color: #fff;
  -webkit-animation: spin 1s ease-in-out infinite;
  animation: spin 1s ease-in-out infinite;
}

.c0 {
  padding-top: 2rem;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <p
        role="status"
      >
        <span
          class="c2"
          width="50"
        />
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<AccountList/> - layout hides spinner on api call finish 1`] = `
<DocumentFragment>
  .c1 {
  margin: 0 auto;
  width: 80%;
  border: 1px solid #dfdfdf;
  border-radius: .25rem;
}

.c2 {
  height: 3rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #f7f7f7;
}

.c3 {
  padding: 1rem;
  display: grid;
}

.c4 {
  display: grid;
  grid-template-columns: repeat(3,1fr);
  overflow: scroll;
}

.c4 span {
  padding: 1rem 0.3rem;
  border-bottom: 1px solid #dfdfdf;
}

.c4 span:nth-child(-n + 3) {
  border-bottom: 1px solid #000;
}

.c4 span:nth-last-child(-n + 3) {
  border-bottom: none;
}

.c0 {
  padding-top: 2rem;
}

@media (min-width:576px) {
  .c2 {
    height: 3.5rem;
  }
}

@media (min-width:576px) {
  .c2 h1 {
    font-size: 1.1rem;
  }
}

@media (min-width:768px) {
  .c2 h1 {
    font-size: 1.2rem;
  }
}

@media (min-width:992px) {
  .c2 h1 {
    font-size: 1.5rem;
  }
}

@media (min-width:768px) {
  .c4 span {
    padding: 1.2rem 0.3rem;
  }
}

@media (min-width:992px) {
  .c4 span {
    font-size: 1.1rem;
  }
}

@media (min-width:768px) {
  .c4 span:nth-child(-n + 3) > * {
    font-size: 1.1rem;
  }
}

@media (min-width:992px) {
  .c4 span:nth-child(-n + 3) > * {
    font-size: 1.3rem;
  }
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <h1>
          Account list
        </h1>
      </div>
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <span
            data-testid="header"
          >
            <strong>
              Name
            </strong>
          </span>
          <span
            data-testid="header"
          >
            <strong>
              Profit & Loss
            </strong>
          </span>
          <span
            data-testid="header"
          >
            <strong>
              Account Type
            </strong>
          </span>
          <span
            data-testid="cell"
          >
            Spread bet
          </span>
          <span
            data-testid="cell"
          >
            $ 0.23
          </span>
          <span
            data-testid="cell"
          >
            Spread bet account
          </span>
          <span
            data-testid="cell"
          >
            New Spread bet
          </span>
          <span
            data-testid="cell"
          >
            $ -679
          </span>
          <span
            data-testid="cell"
          >
            Spread bet account
          </span>
          <span
            data-testid="cell"
          >
            Spread bet - demo
          </span>
          <span
            data-testid="cell"
          >
            $ 16.211
          </span>
          <span
            data-testid="cell"
          >
            Spread bet account
          </span>
          <span
            data-testid="cell"
          >
            CFD - demo
          </span>
          <span
            data-testid="cell"
          >
            € 16679
          </span>
          <span
            data-testid="cell"
          >
            CFD account
          </span>
          <span
            data-testid="cell"
          >
            Stockbroking - DEMO
          </span>
          <span
            data-testid="cell"
          >
            $ 200045
          </span>
          <span
            data-testid="cell"
          >
            Share dealing account
          </span>
          <span
            data-testid="cell"
          >
            Stockbroking
          </span>
          <span
            data-testid="cell"
          >
            € 200045
          </span>
          <span
            data-testid="cell"
          >
            Share dealing account
          </span>
          <span
            data-testid="cell"
          >
            My CFD
          </span>
          <span
            data-testid="cell"
          >
            $ -10
          </span>
          <span
            data-testid="cell"
          >
            CFD account
          </span>
          <span
            data-testid="cell"
          >
            Forex Acccount
          </span>
          <span
            data-testid="cell"
          >
            $ 10
          </span>
          <span
            data-testid="cell"
          >
            Forex account
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<AccountList/> - server error displays error message on accountList fetch failure 1`] = `
<DocumentFragment>
  .c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  text-align: center;
}

.c2 {
  color: #dd3546;
}

.c3 {
  margin: 2rem auto 0 auto;
  padding: 0.5rem;
  width: 50%;
  background-color: #0f7bff;
  border: 1px solid #dfdfdf;
  border-radius: .25rem;
  border: 1px solid transparent;
  color: #fff;
  font-weight: 400;
}

.c0 {
  padding-top: 2rem;
}

@media (min-width:992px) {
  .c2 {
    font-size: 1.2rem;
  }
}

@media (min-width:992px) {
  .c3 {
    font-size: 1.2rem;
  }
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div>
        <span
          class="c2"
        >
          Something went wrong fetching account list
        </span>
        <button
          class="c3"
        >
          Try again
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<AccountList/> - server error displays spinner during refetch api call 1`] = `
<DocumentFragment>
  .c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  text-align: center;
}

.c2 {
  width: 50px;
  display: block;
  aspect-ratio: 1;
  border: 3px solid #0f7bff;
  border-radius: 50%;
  border-top-color: #fff;
  -webkit-animation: spin 1s ease-in-out infinite;
  animation: spin 1s ease-in-out infinite;
}

.c0 {
  padding-top: 2rem;
}

@media (min-width:992px) {

}

@media (min-width:992px) {

}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <p
        role="status"
      >
        <span
          class="c2"
          width="50"
        />
      </p>
    </div>
  </div>
</DocumentFragment>
`;
